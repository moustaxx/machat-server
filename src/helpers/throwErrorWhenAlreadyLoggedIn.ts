import { ApolloError } from 'apollo-server-errors';

function throwErrorWhenAlreadyLoggedIn(isLoggedIn: boolean | null) {
    if (isLoggedIn) {
        throw new ApolloError('You are already logged in!', 'ALREADY_LOGGED_IN');
    }
}

export default throwErrorWhenAlreadyLoggedIn;
