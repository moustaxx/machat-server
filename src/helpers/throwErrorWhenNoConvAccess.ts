import { ForbiddenError } from 'apollo-server-errors';
import { TPrisma } from '../prismaClient';

const throwErrorWhenNoConvAccess = async (
    prisma: TPrisma,
    userID: number,
    conversationID: number,
): Promise<void> => {
    const data = await prisma.person.findUnique({
        select: {
            isAdmin: true,
            conversations: {
                select: {
                    id: true,
                },
            },
        },
        where: { id: userID },
    });

    const isParticipated = !!data?.conversations.find((conv) => {
        return conv.id === conversationID;
    });

    if (!data?.isAdmin && !isParticipated) {
        throw new ForbiddenError('Insufficient permissions');
    }
};

export default throwErrorWhenNoConvAccess;
