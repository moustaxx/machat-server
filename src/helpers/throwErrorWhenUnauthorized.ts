import { ApolloError } from 'apollo-server-errors';

function throwErrorWhenUnauthorized(clientID: number | null): asserts clientID is number {
    if (!clientID) {
        throw new ApolloError('You must be logged in!', 'UNAUTHORIZED');
    }
}

export default throwErrorWhenUnauthorized;
