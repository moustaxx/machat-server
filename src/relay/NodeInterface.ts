import { Field, ID, InterfaceType } from 'type-graphql';
import { ConversationType, MessageType, PersonType } from '../modules';

@InterfaceType({
    isAbstract: true,
    description: 'An object with a global ID.',
    resolveType: (value) => {
        if ('createdAt' in value && 'name' in value) return ConversationType;
        if ('authorID' in value) return MessageType;
        if ('username' in value) return PersonType;
        throw new Error('Unknown type');
    },
})
export abstract class Node {
    @Field(() => ID, {
        name: 'id',
        description: 'The global ID of the object.',
    })
    readonly globalId!: string;
}
