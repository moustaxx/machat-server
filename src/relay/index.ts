export * from './ConnectionArgs';
export * from './ConnectionType';
export * from './NodeInterface';
export * from './NodeResolver';
export * from './PageInfo';
export * from './EdgeType';
