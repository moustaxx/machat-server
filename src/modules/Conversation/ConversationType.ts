import { Connection, ConnectionArguments, findManyCursorConnection } from '@devoxa/prisma-relay-cursor-connection';
import { Person, Message, Prisma } from '@prisma/client';
import { Args, Ctx, Field, FieldResolver, ObjectType, Resolver, Root } from 'type-graphql';

import { Context } from '../../context';
import { ConversationLastReadArgs } from '../../generated/type-graphql';
import cursorUtils from '../../helpers/cursor';
import { ConnectionArgs, ConnectionType, EdgeType, Node } from '../../relay';
import { LastReadType } from '../LastRead';
import { MessageConnection, MessageType } from '../Message';
import { PersonConnection, PersonType } from '../Person';

@ObjectType({ implements: Node })
export class ConversationType extends Node {
    id!: number;

    @Field((_type) => Date)
    createdAt!: Date;

    @Field((_type) => String)
    name!: string;

    @Field((_type) => PersonConnection)
    participants!: Connection<PersonType>;

    @Field((_type) => MessageConnection)
    messages!: MessageType[];

    lastRead!: LastReadType[];
}

@Resolver((_of) => ConversationType)
export class ConversationTypeResolver {
    @FieldResolver((_type) => [PersonConnection])
    async participants(
    // eslint-disable-next-line @typescript-eslint/indent
        @Root() conversation: ConversationType,
        @Ctx() { prisma }: Context,
        @Args((_type) => ConnectionArgs) args: ConnectionArguments,
    ) {
        const where = {
            conversations: {
                some: { id: conversation.id },
            },
        };
        return findManyCursorConnection<Person, Prisma.PersonWhereUniqueInput>(
            async (convArgs) => prisma.person.findMany({ ...convArgs, where }),
            async () => prisma.person.count({ where }),
            args,
            cursorUtils,
        );
    }

    @FieldResolver((_type) => [MessageConnection])
    async messages(
    // eslint-disable-next-line @typescript-eslint/indent
        @Root() conversation: ConversationType,
        @Ctx() { prisma }: Context,
        @Args((_type) => ConnectionArgs) args: ConnectionArguments,
    ) {
        const where = {
            conversation: { id: conversation.id },
        };
        return findManyCursorConnection<Message, Prisma.MessageWhereUniqueInput>(
            async (convArgs) => prisma.message.findMany({ ...convArgs, where }),
            async () => prisma.message.count({ where }),
            args,
            cursorUtils,
        );
    }

    @FieldResolver((_type) => [LastReadType])
    async lastRead(
    // eslint-disable-next-line @typescript-eslint/indent
        @Root() conversation: ConversationType,
        @Ctx() { prisma }: Context,
        @Args() args: ConversationLastReadArgs,
    ) {
        return prisma.conversation.findUnique({
            where: { id: conversation.id },
        }).lastRead(args);
    }
}

@ObjectType()
export class ConversationEdge extends EdgeType(ConversationType) { }

@ObjectType()
export class ConversationConnection extends ConnectionType(ConversationEdge) { }
