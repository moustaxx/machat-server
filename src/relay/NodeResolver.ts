/* eslint-disable @typescript-eslint/indent */
import { Arg, Ctx, FieldResolver, ID, Info, Query, Resolver, Root } from 'type-graphql';

import { Node } from './NodeInterface';
import { Connection } from './ConnectionType';
import { Context } from '../context';
import { PersonResolver, ConversationResolver } from '../modules';
import throwErrorWhenUnauthorized from '../helpers/throwErrorWhenUnauthorized';

export type TNodeModel<TModel> = Omit<TModel, 'id' | 'globalId'> & {
    id: `${string}:${string}`;
};

export type TNodeConnection<TModel> = Connection<TNodeModel<TModel>>;

export const toGlobalId = (typename: string, id: string | number) => {
    return `${typename}:${id}`;
};

export const fromGlobalId = (nodeID: string) => {
    const delimiterPos = nodeID.indexOf(':');
    const id = Number(nodeID.substring(delimiterPos + 1));
    const typename = nodeID.substring(0, delimiterPos);

    return { id, typename };
};

@Resolver(() => Node)
export class NodeResolver {
    @FieldResolver()
    globalId(
        @Root() { id }: { id: string | number },
        @Info() { parentType: { name } }: { parentType: { name: string } },
    ): string {
        return toGlobalId(name, id);
    }

    private async fetcher(globalId: string, ctx: Context | Context<true>) {
        throwErrorWhenUnauthorized(ctx.session);
        const { typename, id } = fromGlobalId(globalId);

        switch (typename) {
            case 'ConversationType': {
                return ConversationResolver.prototype.conversation(
                    { whereId: id },
                    ctx as Context<true>,
                );
            }
            case 'PersonType': {
                return PersonResolver.prototype.person({ id }, ctx as Context<true>);
            }
            case 'MessageType': {
                throw new Error('Unallowed type');
            }
            default: {
                throw new Error('Unknown type');
            }
        }
    }

    @Query(() => Node, {
        nullable: true,
        description: 'Fetches an object given its global ID.',
    })
    async node(
        @Arg('id', () => ID, { description: 'The global ID of the object.' })
        globalId: string,
        @Ctx() ctx: Context,
    ): ReturnType<NodeResolver['fetcher']> {
        return this.fetcher(globalId, ctx);
    }

    @Query(() => [Node], {
        nullable: 'items',
        description: 'Fetches objects given their global IDs.',
    })
    nodes(
        @Arg('ids', () => [ID], { description: 'The global IDs of the objects.' })
        globalIds: Array<string>,
        @Ctx() ctx: Context,
    ): Array<ReturnType<NodeResolver['fetcher']>> {
        return globalIds.map(async (id) => this.fetcher(id, ctx));
    }
}
