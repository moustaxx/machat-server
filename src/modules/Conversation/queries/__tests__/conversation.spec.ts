import { initTestServer, ITestUtils } from '../../../../tests/helpers';
import { TNodeModel } from '../../../../relay';
import randomString from '../../../../tests/helpers/randomString';
import { ConversationType } from '../../ConversationType';

let t: ITestUtils;

beforeAll(async () => {
    t = await initTestServer();
});

afterAll(async () => {
    await t.closeTestServer();
});

const queryStringFull = `
    query conversation($whereId: Int!) {
        conversation(whereId: $whereId) {
            id
            name
            createdAt
            participants {
                edges {
                    node {
                        id
                    }
                }
            }
            messages {
                edges {
                    node {
                        id
                    }
                }
            }
        }
    }
`;

const queryString = `
    query conversation($whereId: Int!) {
        conversation(whereId: $whereId) {
            id
        }
    }
`;

it('should return conversation', async () => {
    const { cookies, user } = await t.createRandomUserAndLogin();

    const name = randomString(8);
    const conversation = await t.prisma.conversation.create({
        data: {
            name,
            participants: { connect: { id: user.id } },
        },
    });

    type TData = { conversation: TNodeModel<ConversationType> };
    const { data } = await t.gqlQuery<TData>({
        cookies,
        query: queryStringFull,
        variables: { whereId: conversation.id },
    });

    expect(data).toMatchSnapshot({
        conversation: {
            name: expect.any(String),
            createdAt: expect.any(String),
        },
    });

    expect(data.conversation.name).toEqual(name);
});

it('should throw NO_DATA error when no data', async () => {
    const { cookies } = await t.createRandomUserAndLogin({ isAdmin: true });

    const { errors } = await t.gqlQuery({
        cookies,
        query: queryString,
        variables: { whereId: 999 },
    });

    const errorCode = errors?.[0].extensions?.code;
    expect(errorCode).toEqual('NO_DATA');
});

it('should throw FORBIDDEN error when not permitted', async () => {
    const { cookies } = await t.createRandomUserAndLogin();

    const conversation = await t.prisma.conversation.create({
        data: { name: randomString(8) },
    });

    const { errors } = await t.gqlQuery({
        query: queryString,
        cookies,
        variables: { whereId: conversation.id },
    });

    const errorCode = errors?.[0].extensions?.code;
    expect(errorCode).toEqual('FORBIDDEN');
});
