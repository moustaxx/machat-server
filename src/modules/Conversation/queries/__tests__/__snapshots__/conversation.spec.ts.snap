// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return conversation 1`] = `
Object {
  "conversation": Object {
    "createdAt": Any<String>,
    "id": "ConversationType:1",
    "messages": Object {
      "edges": Array [],
    },
    "name": Any<String>,
    "participants": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "PersonType:1",
          },
        },
      ],
    },
  },
}
`;
